theano/tensor/nlinalg.py

```diff
@@ -179,6 +185,10 @@ class ExtractDiag(Op):
     __props__ = ("view",)
 
     def __init__(self, view=False):
+        warnings.warn("DeprecationWarning: theano.tensor.nlinalg.ExtractDiag"
```

Comment:
ExtractDiag have an optimization in theano/sandbox/cuda/opt.py. Double check that this opt also work for Diagonal.

Make sure that Diag and Diagonal have the grad and the infershape implemented.